### ISSUE TITLE
Hyperparameter Tuning

### ISSUE THREAD

**Issue by upadhyan** [benibaeumle/Learning-Shapelets#2]
Hi! I had a question about your HP Tuning process:

> we performed a hyperparameter search over 18 different shapelet configurations. For each of the 18 shapelet configurations, we trained three models

How do you select the best configuration? Do you utilize a fixed validation set or cross-validation?

**Comment by benibaeumle** [comment]
Hi,

the UCR archive uses a fixed train/test split. I just stuck to this default split and each configuration was trained and evaluated using this (same) split. So, a fixed validation set.

Here is the relevant passage from the UCR archive for why they chose to do split the UCR archive like this (Chapter 3, V):

> Many researchers, especially those coming from a traditional machine learning background have criticized the archive for having a single train/test split. The original motivation for fixing the train and test set was to allow exact reproducibility. Suppose we simply suggested doing five-fold cross validation. Further suppose, someone claimed to be able to achieve an accuracy of A, on some data sets in the archive. If someone else re-implemented their algorithm and got an accuracy that is slightly lower than A during their five-fold cross validation, it would be difficult to know if that was within the expected variance of different folds, or the result of a bug or a misunderstanding in the new implementation. This issue would be less of a problem if everyone shared their code, and/or had very explicit algorithm descriptions. However, while the culture of open source code is growing in the community, such openness was not always the norm. 
With a single train/test split, and a deterministic algorithm such as 1 NN, failure to exactly reproduce someone else‘s result immediately suggests an issue that should be investigated before proceeding with research. Note that while performing experiments on the single train/test split was always suggested as an absolute minimum sanity check; it did/does not preclude pooling the two splits and then performing K-fold cross validation or any other more rigorous evaluation.

As the last paragraph suggests, you could further divide the train and test sets. Some of the datasets are already pretty small, and I can't remember reading much about further splitting up the datasets in other technique papers (but I might be wrong). So, I just kept it simple.

**Comment by benibaeumle** [comment]
This implementation was part of my master's thesis. Also, I implemented a little research prototype for a visualization system to analyze the decision process of the learning shapelets models. I recently re-deployed the prototype at: https://shapelets.byteskate.de/.

It has a few views, and you can interactively explore the models. I'm sorry if it is not super intuitive (without any explanation). If you want to know something more, don't hesitate to ask me.

I have also sent you a copy of my master's thesis via email (which I found on your GitHub profile) in case you want to know anything in more detail.

**Comment by upadhyan** [comment]
Hi Benedict! This helps so much. I've been trying for ages to recreate the LTS results and I kept getting confounded by the exact issue you brought up (the small training set)!
